examples/multigpu_advtrain/model.py

```diff
@@ -0,0 +1,339 @@
+        else:
+            with tf.variable_scope(self.name):
+                self.set_input_shape_ngpu(x.shape[1:], **kwargs)
+                return self.fprop_noscope(x, **kwargs)
+
+
+class Linear(Layer):
+
+    def __init__(self, num_hid, w_name='W', **kwargs):
+        super(Linear, self).__init__(**kwargs)
+        self.num_hid = num_hid
+        self.input_shape = None
+        self.w_name = w_name
+
+    def set_input_shape(self, input_shape, **kwargs):
+        batch_size, dim = input_shape
+        self.input_shape = [batch_size, dim]
+        self.output_shape = [batch_size, self.num_hid]
+        with tf.variable_scope(self.__class__.__name__):
+            shape = [dim, self.num_hid]
+            init = tf.truncated_normal(shape, stddev=0.1)
+            self.W = self.get_variable(self.w_name, init)
+            self.b = self.get_variable('b', .1 + np.zeros(
+                (self.num_hid,)).astype('float32'))
+
+    def fprop_noscope(self, x, **kwargs):
+        return tf.matmul(x, self.W) + self.b
+
+
+class Conv2D(Layer):
+
+    def __init__(self, output_channels, kernel_shape, strides, padding,
+                 w_name='kernels', *args, **kwargs):
+        super(Conv2D, self).__init__(*args, **kwargs)
+        self.__dict__.update(locals())
+        del self.self
+        self.input_shape = None
+        self.w_name = w_name
+
+    def set_input_shape(self, input_shape, **kwargs):
+        batch_size, rows, cols, input_channels = input_shape
+        kernel_shape = tuple(self.kernel_shape) + (input_channels,
+                                                   self.output_channels)
+        assert len(kernel_shape) == 4
+        assert all(isinstance(e, int) for e in kernel_shape), kernel_shape
+        with tf.variable_scope(self.__class__.__name__):
+            init = tf.truncated_normal(kernel_shape, stddev=0.1)
+            self.kernels = self.get_variable(self.w_name, init)
+            self.b = self.get_variable(
+                'b', .1 + np.zeros((self.output_channels,)).astype('float32'))
+        input_shape = list(input_shape)
+        self.input_shape = input_shape
+        input_shape[0] = 1
+        dummy_batch = tf.zeros(input_shape)
+        dummy_output = self.fprop(dummy_batch)
+        output_shape = [int(e) for e in dummy_output.get_shape()]
+        output_shape[0] = 1
+        self.output_shape = tuple(output_shape)
+
+    def fprop_noscope(self, x, **kwargs):
+        return tf.nn.conv2d(x, self.kernels, (1,) + tuple(self.strides) +
+                            (1,), self.padding) + self.b
+
+
+class ReLU(Layer):
+
+    def __init__(self, **kwargs):
+        super(ReLU, self).__init__(**kwargs)
+
+    def set_input_shape(self, shape, **kwargs):
+        self.input_shape = shape
+        self.output_shape = shape
+
+    def get_output_shape(self):
+        return self.output_shape
+
+    def fprop_noscope(self, x, **kwargs):
+        return tf.nn.relu(x)
+
+
+class Softmax(Layer):
```

Comment:
similarly, how hard would it be to use the one from the tutorials?

Reply:
Removed in 714afc0.